File: src/services/plugin/storage.rs
//! Plugin Storage Protocol Types
//!
//! Defines the JSON-RPC request/response types for plugin storage operations.
//! Plugins use these methods to persist per-user data like taste profiles,
//! sync state, and cached recommendations.
//!
//! ## Architecture
//!
//! Storage is scoped per user-plugin instance. Plugins only specify a key;
//! the host resolves the user_plugin_id from the connection context.
//! This provides architectural isolation - plugins cannot address other
//! plugins' or users' data.
//!
//! ## Methods
//!
//! - `storage/get` - Get a value by key
//! - `storage/set` - Set a value (upsert) with optional TTL
//! - `storage/delete` - Delete a value by key
//! - `storage/list` - List all keys
//! - `storage/clear` - Clear all data

use serde::{Deserialize, Serialize};
use serde_json::Value;

// =============================================================================
// Storage Request Types
// =============================================================================

/// Parameters for `storage/get` method
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct StorageGetRequest {
    /// Storage key to retrieve
    pub key: String,
}

/// Parameters for `storage/set` method
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct StorageSetRequest {
    /// Storage key
    pub key: String,
    /// JSON data to store
    pub data: Value,
    /// Optional expiration timestamp (ISO 8601)
    /// If set, the data will be automatically cleaned up after this time
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub expires_at: Option<String>,
}

/// Parameters for `storage/delete` method
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct StorageDeleteRequest {
    /// Storage key to delete
    pub key: String,
}

// Note: `storage/list` and `storage/clear` take no parameters

// =============================================================================
// Storage Response Types
// =============================================================================

/// Response from `storage/get` method
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct StorageGetResponse {
    /// The stored data, or null if key doesn't exist
    pub data: Option<Value>,
    /// Expiration timestamp (ISO 8601) if TTL was set
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub expires_at: Option<String>,
}

/// Response from `storage/set` method
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct StorageSetResponse {
    /// Always true on success
    pub success: bool,
}

/// Response from `storage/delete` method
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct StorageDeleteResponse {
    /// Whether the key existed and was deleted
    pub deleted: bool,
}

/// Individual key entry in `storage/list` response
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct StorageKeyEntry {
    /// Storage key name
    pub key: String,
    /// Expiration timestamp (ISO 8601) if TTL was set
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub expires_at: Option<String>,
    /// Last update timestamp (ISO 8601)
    pub updated_at: String,
}

/// Response from `storage/list` method
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct StorageListResponse {
    /// All keys for this plugin instance (excluding expired)
    pub keys: Vec<StorageKeyEntry>,
}

/// Response from `storage/clear` method
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct StorageClearResponse {
    /// Number of entries deleted
    pub deleted_count: u64,
}

// =============================================================================
// Permission Check
// =============================================================================

/// Check if a method name is a storage method
pub fn is_storage_method(method: &str) -> bool {
    matches!(
        method,
        "storage/get" | "storage/set" | "storage/delete" | "storage/list" | "storage/clear"
    )
}

// =============================================================================
// Tests
// =============================================================================

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;

    #[test]
    fn test_storage_get_request_serialization() {
        let req = StorageGetRequest {
            key: "taste_profile".to_string(),
        };
        let json = serde_json::to_value(&req).unwrap();
        assert_eq!(json["key"], "taste_profile");
    }

    #[test]
    fn test_storage_get_request_deserialization() {
        let json = json!({"key": "sync_state"});
        let req: StorageGetRequest = serde_json::from_value(json).unwrap();
        assert_eq!(req.key, "sync_state");
    }

    #[test]
    fn test_storage_set_request_serialization() {
        let req = StorageSetRequest {
            key: "recommendations".to_string(),
            data: json!({"items": [1, 2, 3], "score": 0.95}),
            expires_at: Some("2026-02-07T00:00:00Z".to_string()),
        };
        let json = serde_json::to_value(&req).unwrap();
        assert_eq!(json["key"], "recommendations");
        assert_eq!(json["data"]["score"], 0.95);
        assert_eq!(json["expiresAt"], "2026-02-07T00:00:00Z");
    }

    #[test]
    fn test_storage_set_request_without_ttl() {
        let req = StorageSetRequest {
            key: "profile".to_string(),
            data: json!({"version": 1}),
            expires_at: None,
        };
        let json = serde_json::to_value(&req).unwrap();
        assert!(!json.as_object().unwrap().contains_key("expiresAt"));
    }

    #[test]
    fn test_storage_set_request_deserialization_with_ttl() {
        let json = json!({
            "key": "cache",
            "data": [1, 2, 3],
            "expiresAt": "2026-03-01T12:00:00Z"
        });
        let req: StorageSetRequest = serde_json::from_value(json).unwrap();
        assert_eq!(req.key, "cache");
        assert_eq!(req.data, json!([1, 2, 3]));
        assert_eq!(req.expires_at.unwrap(), "2026-03-01T12:00:00Z");
    }

    #[test]
    fn test_storage_set_request_deserialization_without_ttl() {
        let json = json!({
            "key": "state",
            "data": {"cursor": "abc123"}
        });
        let req: StorageSetRequest = serde_json::from_value(json).unwrap();
        assert!(req.expires_at.is_none());
    }

    #[test]
    fn test_storage_delete_request_serialization() {
        let req = StorageDeleteRequest {
            key: "old_cache".to_string(),
        };
        let json = serde_json::to_value(&req).unwrap();
        assert_eq!(json["key"], "old_cache");
    }

    #[test]
    fn test_storage_get_response_with_data() {
        let resp = StorageGetResponse {
            data: Some(json!({"version": 2, "items": []})),
            expires_at: Some("2026-02-08T00:00:00Z".to_string()),
        };
        let json = serde_json::to_value(&resp).unwrap();
        assert_eq!(json["data"]["version"], 2);
        assert_eq!(json["expiresAt"], "2026-02-08T00:00:00Z");
    }

    #[test]
    fn test_storage_get_response_null_data() {
        let resp = StorageGetResponse {
            data: None,
            expires_at: None,
        };
        let json = serde_json::to_value(&resp).unwrap();
        assert!(json["data"].is_null());
        assert!(!json.as_object().unwrap().contains_key("expiresAt"));
    }

    #[test]
    fn test_storage_set_response() {
        let resp = StorageSetResponse { success: true };
        let json = serde_json::to_value(&resp).unwrap();
        assert!(json["success"].as_bool().unwrap());
    }

    #[test]
    fn test_storage_delete_response() {
        let resp = StorageDeleteResponse { deleted: true };
        let json = serde_json::to_value(&resp).unwrap();
        assert!(json["deleted"].as_bool().unwrap());

        let resp2 = StorageDeleteResponse { deleted: false };
        let json2 = serde_json::to_value(&resp2).unwrap();
        assert!(!json2["deleted"].as_bool().unwrap());
    }

    #[test]
    fn test_storage_list_response() {
        let resp = StorageListResponse {
            keys: vec![
                StorageKeyEntry {
                    key: "profile".to_string(),
                    expires_at: None,
                    updated_at: "2026-02-06T10:00:00Z".to_string(),
                },
                StorageKeyEntry {
                    key: "cache".to_string(),
                    expires_at: Some("2026-02-07T00:00:00Z".to_string()),
                    updated_at: "2026-02-06T11:00:00Z".to_string(),
                },
            ],
        };
        let json = serde_json::to_value(&resp).unwrap();
        assert_eq!(json["keys"].as_array().unwrap().len(), 2);
        assert_eq!(json["keys"][0]["key"], "profile");
        assert!(
            !json["keys"][0]
                .as_object()
                .unwrap()
                .contains_key("expiresAt")
        );
        assert_eq!(json["keys"][1]["expiresAt"], "2026-02-07T00:00:00Z");
    }

    #[test]
    fn test_storage_clear_response() {
        let resp = StorageClearResponse { deleted_count: 5 };
        let json = serde_json::to_value(&resp).unwrap();
        assert_eq!(json["deletedCount"], 5);
    }

    #[test]
    fn test_is_storage_method() {
        assert!(is_storage_method("storage/get"));
        assert!(is_storage_method("storage/set"));
        assert!(is_storage_method("storage/delete"));
        assert!(is_storage_method("storage/list"));
        assert!(is_storage_method("storage/clear"));
        assert!(!is_storage_method("metadata/series/search"));
        assert!(!is_storage_method("initialize"));
        assert!(!is_storage_method("storage/unknown"));
    }

    #[test]
    fn test_storage_get_response_deserialization() {
        let json = json!({
            "data": {"genres": ["action", "drama"]},
            "expiresAt": "2026-12-31T23:59:59Z"
        });
        let resp: StorageGetResponse = serde_json::from_value(json).unwrap();
        assert!(resp.data.is_some());
        assert_eq!(resp.expires_at.unwrap(), "2026-12-31T23:59:59Z");
    }

    #[test]
    fn test_storage_key_entry_serialization() {
        let entry = StorageKeyEntry {
            key: "test_key".to_string(),
            expires_at: None,
            updated_at: "2026-02-06T12:00:00Z".to_string(),
        };
        let json = serde_json::to_value(&entry).unwrap();
        assert_eq!(json["key"], "test_key");
        assert_eq!(json["updatedAt"], "2026-02-06T12:00:00Z");
        assert!(!json.as_object().unwrap().contains_key("expiresAt"));
    }
}

File: plugins/sdk-typescript/src/storage.ts
/**
 * Plugin Storage - key-value storage for per-user plugin data
 *
 * Storage is scoped per user-plugin instance. Plugins only specify a key;
 * the host resolves the user context from the connection.
 *
 * Plugins send storage requests as JSON-RPC calls to the host over stdout
 * and receive responses on stdin. This is the reverse of the normal
 * host-to-plugin request flow.
 *
 * @example
 * ```typescript
 * import { PluginStorage } from "@ashdev/codex-plugin-sdk";
 *
 * const storage = new PluginStorage();
 *
 * // Store data
 * await storage.set("taste_profile", { genres: ["action", "drama"] });
 *
 * // Retrieve data
 * const data = await storage.get("taste_profile");
 *
 * // Store with TTL (expires in 24 hours)
 * const expires = new Date(Date.now() + 24 * 60 * 60 * 1000).toISOString();
 * await storage.set("cache", { items: [1, 2, 3] }, expires);
 *
 * // List all keys
 * const keys = await storage.list();
 *
 * // Delete a key
 * await storage.delete("cache");
 *
 * // Clear all data
 * await storage.clear();
 * ```
 */

import type { JsonRpcError, JsonRpcRequest } from "./types/rpc.js";

// =============================================================================
// Storage Types
// =============================================================================

/** Response from storage/get */
export interface StorageGetResponse {
  /** The stored data, or null if key doesn't exist */
  data: unknown | null;
  /** Expiration timestamp (ISO 8601) if TTL was set */
  expiresAt?: string;
}

/** Response from storage/set */
export interface StorageSetResponse {
  /** Always true on success */
  success: boolean;
}

/** Response from storage/delete */
export interface StorageDeleteResponse {
  /** Whether the key existed and was deleted */
  deleted: boolean;
}

/** Individual key entry from storage/list */
export interface StorageKeyEntry {
  /** Storage key name */
  key: string;
  /** Expiration timestamp (ISO 8601) if TTL was set */
  expiresAt?: string;
  /** Last update timestamp (ISO 8601) */
  updatedAt: string;
}

/** Response from storage/list */
export interface StorageListResponse {
  /** All keys for this plugin instance (excluding expired) */
  keys: StorageKeyEntry[];
}

/** Response from storage/clear */
export interface StorageClearResponse {
  /** Number of entries deleted */
  deletedCount: number;
}

// =============================================================================
// Storage Error
// =============================================================================

/** Error from a storage operation */
export class StorageError extends Error {
  constructor(
    message: string,
    public readonly code: number,
    public readonly data?: unknown,
  ) {
    super(message);
    this.name = "StorageError";
  }
}

// =============================================================================
// Plugin Storage Client
// =============================================================================

/** Write function signature for sending JSON-RPC requests */
type WriteFn = (line: string) => void;

/**
 * Client for plugin key-value storage.
 *
 * Sends JSON-RPC requests to the host process over stdout and reads
 * responses on stdin. Each request gets a unique ID so responses can
 * be correlated even if they arrive out of order.
 */
export class PluginStorage {
  private nextId = 1;
  private pendingRequests = new Map<
    string | number,
    {
      resolve: (value: unknown) => void;
      reject: (error: Error) => void;
    }
  >();
  private writeFn: WriteFn;

  /**
   * Create a new storage client.
   *
   * @param writeFn - Optional custom write function (defaults to process.stdout.write).
   *                  Useful for testing or custom transport layers.
   */
  constructor(writeFn?: WriteFn) {
    this.writeFn =
      writeFn ??
      ((line: string) => {
        process.stdout.write(line);
      });
  }

  /**
   * Get a value by key
   *
   * @param key - Storage key to retrieve
   * @returns The stored data and optional expiration, or null data if key doesn't exist
   */
  async get(key: string): Promise<StorageGetResponse> {
    return (await this.sendRequest("storage/get", { key })) as StorageGetResponse;
  }

  /**
   * Set a value by key (upsert - creates or updates)
   *
   * @param key - Storage key
   * @param data - JSON-serializable data to store
   * @param expiresAt - Optional expiration timestamp (ISO 8601)
   * @returns Success indicator
   */
  async set(key: string, data: unknown, expiresAt?: string): Promise<StorageSetResponse> {
    const params: Record<string, unknown> = { key, data };
    if (expiresAt !== undefined) {
      params.expiresAt = expiresAt;
    }
    return (await this.sendRequest("storage/set", params)) as StorageSetResponse;
  }

  /**
   * Delete a value by key
   *
   * @param key - Storage key to delete
   * @returns Whether the key existed and was deleted
   */
  async delete(key: string): Promise<StorageDeleteResponse> {
    return (await this.sendRequest("storage/delete", { key })) as StorageDeleteResponse;
  }

  /**
   * List all keys for this plugin instance (excluding expired)
   *
   * @returns List of key entries with metadata
   */
  async list(): Promise<StorageListResponse> {
    return (await this.sendRequest("storage/list", {})) as StorageListResponse;
  }

  /**
   * Clear all data for this plugin instance
   *
   * @returns Number of entries deleted
   */
  async clear(): Promise<StorageClearResponse> {
    return (await this.sendRequest("storage/clear", {})) as StorageClearResponse;
  }

  /**
   * Handle an incoming JSON-RPC response line from the host.
   *
   * Call this method from your readline handler to deliver responses
   * back to pending storage requests.
   */
  handleResponse(line: string): void {
    const trimmed = line.trim();
    if (!trimmed) return;

    let parsed: unknown;
    try {
      parsed = JSON.parse(trimmed);
    } catch {
      // Not JSON - ignore
      return;
    }

    const obj = parsed as Record<string, unknown>;

    // Only handle responses (have "result" or "error", no "method")
    if (obj.method !== undefined) {
      // This is a host-to-plugin request, not a storage response - ignore
      return;
    }

    const id = obj.id;
    if (id === undefined || id === null) return;

    const pending = this.pendingRequests.get(id as string | number);
    if (!pending) return;

    this.pendingRequests.delete(id as string | number);

    if ("error" in obj && obj.error) {
      const err = obj.error as JsonRpcError;
      pending.reject(new StorageError(err.message, err.code, err.data));
    } else {
      pending.resolve(obj.result);
    }
  }

  /**
   * Cancel all pending requests (e.g. on shutdown).
   */
  cancelAll(): void {
    for (const [, pending] of this.pendingRequests) {
      pending.reject(new StorageError("Storage client stopped", -1));
    }
    this.pendingRequests.clear();
  }

  // ===========================================================================
  // Internal
  // ===========================================================================

  private sendRequest(method: string, params: unknown): Promise<unknown> {
    const id = this.nextId++;

    const request: JsonRpcRequest = {
      jsonrpc: "2.0",
      id,
      method,
      params,
    };

    return new Promise((resolve, reject) => {
      this.pendingRequests.set(id, { resolve, reject });

      try {
        this.writeFn(`${JSON.stringify(request)}\n`);
      } catch (err) {
        this.pendingRequests.delete(id);
        const message = err instanceof Error ? err.message : "Unknown write error";
        reject(new StorageError(`Failed to send request: ${message}`, -1));
      }
    });
  }
}
